### PR TITLE
Support for simultaneous read from local and remote parallel replica

### PR DESCRIPTION
**Pull request by nikitamikhaylov**
### Changelog category (leave one):

- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Now if a shard has local replica we create a local plan and a plan to read from all remote replicas. They have shared initiator which coordinates reading. 

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Interpreters/Cluster.cpp [hunk 1]
```diff
@@ -485,6 +487,7 @@ Cluster::Cluster(const Poco::Util::AbstractConfiguration & config,
             }
 
             Addresses shard_local_addresses;
+            Addresses all_local_addresses;
```
Comment (antonio2368) [inline src/Interpreters/Cluster.cpp:490]: But the above vector stores local addresses right?
Shouldn't this be `shard_all_addresses`?

File: src/Interpreters/ClusterProxy/SelectStreamFactory.cpp [hunk 2]
```diff
@@ -63,18 +65,15 @@ void SelectStreamFactory::createForShard(
 
     auto emplace_local_stream = [&]()
     {
-        local_plans.emplace_back(createLocalPlan(query_ast, header, context, processed_stage, shard_info.shard_num, shard_count, /*coordinator=*/nullptr));
+        local_plans.emplace_back(createLocalPlan(query_ast, header, context, processed_stage, shard_info.shard_num, shard_count, 0, 0, /*coordinator=*/nullptr));
```
Comment (antonio2368) [inline src/Interpreters/ClusterProxy/SelectStreamFactory.cpp:68]: Add comments for hardcoded values like for `coordinator`. One of the builds in CI catch unmatched argument names so I think it's good rule to enforce.

File: src/Interpreters/ClusterProxy/SelectStreamFactory.cpp [hunk 3]
```diff
@@ -173,5 +172,106 @@ void SelectStreamFactory::createForShard(
...
+
+    const auto & settings = context->getSettingsRef();
+
+    auto is_local_replica_obsolete = [&]()
+    {
+        auto resolved_id = context->resolveStorageID(main_table);
+        auto main_table_storage = DatabaseCatalog::instance().tryGetTable(resolved_id, context);
+
+        if (!main_table_storage)
+            throw DB::Exception(ErrorCodes::ALL_REPLICAS_ARE_STALE, "");
```
Comment (KochetovNicolai) [inline src/Interpreters/ClusterProxy/SelectStreamFactory.cpp:199]: Do not understand.
In which situation can we get nullptr storage from catalog? Why it means that all replicas are stale?
Exception message is empty.

**Review by KochetovNicolai — APPROVED** [review]
Generally LGTM, code with parallel replicas more clear now

**Pull request merged**